File: backend/product/typescript/src/exceptions/error-exception.ts
import { setErrorMessages, setInternalServerErrorErrorMessages } from "../helpers/error-message";
import { setResponse } from "../helpers/response-message";
import { ResponseException } from "./response-exception";
// import { setInternalServerErrorErrorMessages } from "./exception";
import { Response } from "express";

export function errorHandler(error: unknown, requestId: string) {
    if (error instanceof Error) {
        console.log(JSON.stringify({logTime: (new Date()).toISOString(), app: "backend-product-typescript", requestId: requestId, error: error.stack}));
        if (error instanceof ResponseException) {
            throw new ResponseException(error.status, error.errorMessages, error.message)
        } else {
            throw new ResponseException(500, setInternalServerErrorErrorMessages(), "internal server error")
        }
    } else {
        throw new ResponseException(500, setInternalServerErrorErrorMessages(), "internal server error")
    }
    // console.log(JSON.stringify({logTime: (new Date()).toISOString(), app: "backend-product-typescript", requestId: requestId, error: error.stack}));
    // if (error instanceof ResponseException) {
    //     throw new ResponseException(error.status, error.message)
    // } else {
    //     throw new ResponseException(500, setInternalServerError())
    // }
}

export function errorHandlerResponse(res: Response, error: unknown, requestId: string): Response {
    let status: number = 0
    let response: any = null
    if (error instanceof Error) {
        if (error instanceof ResponseException) {
            // const response = setResponse(null, error.errorMessages)
            // return res.status(error.status).json(response)
            status = error.status
            response = setResponse(null, error.errorMessages)
        } else {
            // const response = setResponse(null, setInternalServerErrorErrorMessages())
            // return res.status(500).json(response)
            status = 500
            response = setResponse(null, setInternalServerErrorErrorMessages())
        }
    } else {
        // const response = setResponse(null, setInternalServerErrorErrorMessages())
        // return res.status(500).json(response)
        status = 500
        response = setResponse(null, setInternalServerErrorErrorMessages())
    }
    return res.status(status).json(response)
}

File: backend/order/python/commons/exception/error_exception.py
from commons.exception.response_exception import ResponseException
import datetime
import traceback
from flask import jsonify

def error_handler(error, requestId):
    print({"logTime": datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f"), "app": "backend-order-python", "requestId": requestId, "error": traceback.format_exc()})
    if isinstance(error, ResponseException):
        raise ResponseException(error.status, error.error_messages, error.message)
    else:
        raise ResponseException(500, set_internal_server_error(), "internal server error")

def set_error_messages(message):
    return [{"field": "message", "message": message}]

def set_internal_server_error():
    return [{"field": "message", "message": "internal server error"}]

def error_handler_response(error):
    if isinstance(error, ResponseException):
        return jsonify({"data": None, "errors": error.error_messages}), error.status
    else:
        return jsonify({"data": None, "errors": set_internal_server_error()}), 500